src/codegen/barrier.cpp

```diff
@@ -0,0 +1,39 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// barrier.cpp
+//
+// Identification: src/codegen/barrier.cpp
+//
+// Copyright (c) 2015-17, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include "common/logger.h"
+#include "codegen/barrier.h"
+#include "codegen/utils/oa_hash_table.h"
+
+namespace peloton {
+namespace codegen {
+
+void Barrier::InitInstance(Barrier *ins, uint64_t n_workers)
```

Comment:
I'm so concerned about all the pointers flying around. From my point of view, "instantiate", "set", and "destroy" are all very dangerous - you might forget to destroy, or set barrier after the barrier is set. This causes trouble in debugging, too. But again, I believe you have your own reason in designing all these functions.